File: src/SimpleLevelEditor.Formats.Types/Level/EntityPropertyValue.fs
namespace SimpleLevelEditor.Formats.Types.Level

open System.Numerics
open SimpleLevelEditor.Formats.Types

type EntityPropertyValue =
    | Bool    of Value: bool
    | Int     of Value: int
    | Float   of Value: float32
    | Vector2 of Value: Vector2
    | Vector3 of Value: Vector3
    | Vector4 of Value: Vector4
    | String  of Value: string
    | Rgb     of Value: Rgb
    | Rgba    of Value: Rgba

    member this.DeepCopy() =
        match this with
        | Bool    v -> Bool v
        | Int     v -> Int v
        | Float   v -> Float v
        | Vector2 v -> Vector2 v
        | Vector3 v -> Vector3 v
        | Vector4 v -> Vector4 v
        | String  v -> String v
        | Rgb     v -> Rgb v
        | Rgba    v -> Rgba v

    member this.ToDisplayString() =
        match this with
        | Bool    v -> v.ToDisplayString
        | Int     v -> v.ToDisplayString
        | Float   v -> v.ToDisplayString
        | Vector2 v -> v.ToDisplayString
        | Vector3 v -> v.ToDisplayString
        | Vector4 v -> v.ToDisplayString
        | String  v -> v
        | Rgb     v -> v.ToDisplayString()
        | Rgba    v -> v.ToDisplayString()

File: src/SimpleLevelEditor.Formats.Level/EntityPropertyValue.cs
using Dunet;
using SimpleLevelEditor.Formats.Core;
using System.Diagnostics;

namespace SimpleLevelEditor.Formats.Level;

[Union]
public partial record EntityPropertyValue
{
	// TODO: Use a better naming convention:
	// Bool
	// I32
	// F32
	// Vec2
	// Vec3
	// Vec4
	// Str
	// Rgba
	// Rgb
	public sealed partial record Bool(bool Value);
	public sealed partial record Int(int Value);
	public sealed partial record Float(float Value);
	public sealed partial record Vector2(System.Numerics.Vector2 Value);
	public sealed partial record Vector3(System.Numerics.Vector3 Value);
	public sealed partial record Vector4(System.Numerics.Vector4 Value);
	public sealed partial record String(string Value);
	public sealed partial record Rgba(Formats.Core.Rgba Value);
	public sealed partial record Rgb(Formats.Core.Rgb Value);

	public EntityPropertyValue DeepCopy()
	{
		return this switch
		{
			Bool b => new Bool(b.Value),
			Int i => new Int(i.Value),
			Float f => new Float(f.Value),
			Vector2 v2 => new Vector2(v2.Value),
			Vector3 v3 => new Vector3(v3.Value),
			Vector4 v4 => new Vector4(v4.Value),
			String s => new String(s.Value),
			Rgba rgba => new Rgba(rgba.Value),
			Rgb rgb => new Rgb(rgb.Value),
			_ => throw new UnreachableException(),
		};
	}

	public string ToDisplayString()
	{
		return this switch
		{
			Bool b => b.Value.ToDisplayString(),
			Int i => i.Value.ToDisplayString(),
			Float f => f.Value.ToDisplayString(),
			Vector2 v2 => v2.Value.ToDisplayString(),
			Vector3 v3 => v3.Value.ToDisplayString(),
			Vector4 v4 => v4.Value.ToDisplayString(),
			String s => s.Value,
			Rgba rgba => rgba.Value.ToDisplayString(),
			Rgb rgb => rgb.Value.ToDisplayString(),
			_ => throw new UnreachableException(),
		};
	}
}
